### ISSUE TITLE
Failed to initialize FTDI SPI

### ISSUE THREAD

**Issue by romansil** [vbendeb/tpm2_server#1]
Hello,

I have problems with getting the tpm2_server running with a hardware TPM. The simulator works fine.
My setup:
- Infineon SLB 9670 TPM2.0
- FTDI Calbe wih a FT232H chip
- tpm2_server (24b103bb7cc8466fba255f689f1fe07cda684e82)

My output of starting `./ntpm -dd`:

```
Opening socket on port 9883
Starting MPSSE at 1000 kHz

unknown did_vid: 0xffffffff
Failed to initialize FTDI SPI
R d40f00: ffffffff
```

Does anyone know this problem?

Kind regards,
Roman

**Comment by vbendeb** [comment]
This shows a problem when there is no data transferred from the TPM to the
controller on the MISO line (it is all ones).

Not sure what the root cause is though, you'd need to probe the connection
to determine what exactly is wrong.

HTH,
-vb

**Comment by romansil** [comment]
@vbendeb thank you for your reply. I did some experiments using libmpsse (with python):
- connecting MISO and MOSI gives the same result as sent.
- sending things like `80 d4 0f 00` lead to a response of `00 00 00 1F` or `00 00 00 07` (I can tell you the exact in and output tomorrow, when I'm in my office again). After the 4 Bytes, I only receive 0xFF.

For me it's hard to understand the SPI hardware protocol. Is there another documentation than the "TCG PC Client Specific TPM Interface Specification"?

**Comment by vbendeb** [comment]
Roman, SPI specification is well described in many places, here for
instance: https://en.wikipedia.org/wiki/Serial_Peripheral_Interface_Bus

The PC client Interface Specification describes the protocol  the TPM runs
on top of SPI.

when you're saying '80 d4 0f 00 lead to a response of 00 00 00 1F' you mean
8 bytes total were clocked and the response was in the last 4 bytes - it
makes sense if so.

Note that this server implementation has been used with a proprietary tpm
implementation, which is not necessarily fully compliant with the TCG spec.

Do you have some protocol analyzer connected to the setup?

-vb

**Comment by romansil** [comment]
vb, I'm familiar with the SPI bus.

At the moment I trusted the libmpssi to do the correct thing. Today I will verify this with a measurement.

> when you're saying '80 d4 0f 00 lead to a response of 00 00 00 1F' you mean 8 bytes total were clocked and the response was in the last 4 bytes - it makes sense if so.
I used the `Transfer` function to send '80 d4 0f 00` and this receives `00 00 00 1F'. If using `Read` afterwards `FF FF FF ...` is received.

My source:
```python
from __future__ import print_function
from time import sleep
from mpsse import *
from hexdump import hexdump


m = MPSSE(SPI0, TWELVE_MHZ, MSB)

m.PinLow(GPIOL0)
sleep(0.1)
m.PinHigh(GPIOL0)

m.Start()

msg = '\x85\x00\x0F\xD4'
print('msg:')
hexdump(msg)
rsp = m.Transfer(msg)
print('response Transfer:')
hexdump(rsp)

rsp = m.Read(16)
print('response Read:')
hexdump(rsp)

m.Stop()
m.Close()
```

And the output:
```
msg:
00000000: 85 00 0F D4                                       ....
response Transfer:
00000000: 00 00 00 0F                                       ....
response Read:
00000000: FF FF FF FF FF FF FF FF  FF FF FF FF FF FF FF FF  ................
```

When appending zero or 0xFF bytes to `msg`, the response Transfer also gets the 0xFF bytes in.
```python
msg1 = '\x85\x00\x0F\xD4\0\0\0\0' 
msg2 = '\x85\x00\x0F\xD4\xFF\xFF\xFF\xFF'
```
Output:
```
response Transfer:
00000000: 00 00 00 0F FF FF FF FF                           ........
```

But when transfering only zero bytes after the reset, the TPM answers a little bit more:
```
msg:
00000000: 00 00 00 00 00 00 00 00  00 00 00 00 00 00 00 00  ................
00000010: 00 00 00 00 00 00 00 00  00 00 00 00 00 00 00 00  ................
00000020: 00 00 00 00 00 00 00 00  00 00 00 00 00 00 00 00  ................
00000030: 00 00 00 00 00 00 00 00  00 00 00 00 00 00 00 00  ................
response Transfer:
00000000: 00 00 00 01 00 00 00 00  02 00 00 00 00 04 00 00  ................
00000010: 00 00 08 00 00 00 00 10  00 00 00 00 20 00 00 00  ............ ...
00000020: 00 40 00 00 00 00 80 00  00 00 01 00 00 00 00 02  .@..............
00000030: 00 00 00 00 04 00 00 00  00 08 00 00 00 00 10 00  ................
```

**Comment by romansil** [comment]
I think there is an electrical problem. When touching the SCK line with my finger, the output changes.
It's still not the expected vendor ID, but I need to fix that first. Maybe the FTDI is not as good as expected.

**Comment by vbendeb** [comment]
FTDI cable I used was pretty reliable. Could it be that the chip and the
cable are for different  voltages (1.8V vs 3.3V).

**Comment by romansil** [comment]
Good idea, but this should work:
- C232HM-DDHSL-0 cable with 3.3V levels and 3.3V supply
- Iridium 9670 - TPM2.0 SPI evaluation board, that's specified for 1.8V and 3.3V

I will try to add pull-up or pull-down resistors next. Or I try to use it with another SPI controller. Thank you for your help so far!

Do you know, if there is a setting for an internal pull up/down or different output stages like push pull or open collector for the FTDI cable?

**Comment by romansil** [comment]
I was able to isolate the problem: Since the C232HM-DDHSL-0 comes with 50cm cable, and I attached 15cm ribbon cable for connecting my TPM, there exist many parallel lines. Because of fast slew rates crosstalk could cause interference pulses on the SCK line. By adding a 1K resistor right beside the TPM I was able to run the tpm2_server.

I can confirm, that the tpm2_server is compartible with the Infineon SLB 9670 TPM2.0. Using some small python scripts as glue logic, I was able to run the Intel [tpm2-tools](https://github.com/intel/tpm2-tools) (v2.0.0) against the tpm2_server. So far I tried `tpm2_startup`, `tpm2_getrandom` and `tpm2_nvlist` successfully.

Thanks vb for your help!
